### PR TITLE
Pdf print page with Chrome DevTools

### PR DESCRIPTION
**Pull request by Maxgit3**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

#6120 Draft of proposed addition to resolve issue

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

set as a draft to see if this fits the proposed addition.

### CODE REVIEW

File: packages/devtools/src/commands/printPage.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import DevToolsDriver from '../devtoolsdriver'
+
+export default async function printPage(options) {
+    const page = DevToolsDriver.getPageHandle()
+    if (!options) {
+        return page.pdf()
+    }
+
+    return page.pdf({
+        landscaper: options.orientation,
```
Comment (christian-bromann) [inline packages/devtools/src/commands/printPage.js:10]:
```suggestion
        landscape: options.orientation === 'landscape',
```

File: packages/devtools/src/commands/printPage.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import DevToolsDriver from '../devtoolsdriver'
+
+export default async function printPage(options) {
+    const page = DevToolsDriver.getPageHandle()
```
Comment (christian-bromann) [inline packages/devtools/src/commands/printPage.js:4]:
```suggestion
    const page = this.getPageHandle()
```

File: packages/devtools/src/commands/printPage.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+            top: options.top,
+            right: options.right,
+            bottom: options.bottom,
+            left: options.left
+        },
+        preferCSSPageSize: !options.shrinkToFit,
+        pageRanges: options.pageRanges
+    })
+
+}
```
Comment (christian-bromann) [inline packages/devtools/src/commands/printPage.js:25]:
```suggestion
}

```

**Review by christian-bromann — CHANGES_REQUESTED** [review]
Thank you for taking this on, a couple of comments:

- let's use TypeScript like in all other command files
- let's add a protocol test in `/e2e/protocol.test.ts
- add some comments to the top of the file (see other devtools commands)

**Comment by christian-bromann** [comment]
ping @Maxgit3 any help I can provide on this?

**Comment by Maxgit3** [comment]
sorry about the delayed response.  I am looking over creating an e2e test.  I am getting a buffer instead of a string.

**Comment by Maxgit3** [comment]
should I save the pdf to a temp location and then confirm if the data is a string?

**Comment by christian-bromann** [comment]
@Maxgit3 just checking if the return value is a string should be enough.

**Comment by Maxgit3** [comment]
updated progress.   looked through the puppeteer documents.  don't see anything about puppeteer returning a string for pdf file.

File: packages/devtools/src/commands/printPage.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import DevToolsDriver from '../devtoolsdriver'
+
+export default async function printPage(this: DevToolsDriver, options: any) {
+    const page = this.getPageHandle()
+    if (!options) {
+        return page.pdf()
+    }
+
+    return page.pdf({
```
Comment (christian-bromann) [inline packages/devtools/src/commands/printPage.ts:17]: In order for it to return a string an be compliant with the WebDriver protocol I suggest:

```js
const pdfBuffer = await page.pdf({...})
return pdfBuffer.toString()
```

**Review by christian-bromann — COMMENTED** [review]
see my comment, happy to assist if anything is unclear

**Comment by Maxgit3** [comment]
Sorry about the delay.  I will update

File: e2e/protocol.test.js [hunk 5]
```diff
@@ -335,6 +335,32 @@ describe('executeScript', () => {
...
+            bottom: '1in',
+            left: '1in'
+        },
+        preferCSSPageSize: false,
+        pageRanges: ''
+    }
+
+    it('expect string', async () => {
+        const stringData = await browser.printPage(opt)
+        expect(stringData).toBe('string')
```
Comment (christian-bromann) [inline e2e/protocol.test.js:360]:
```suggestion
        expect(typeof stringData).toBe('string')
```

**Review by christian-bromann — CHANGES_REQUESTED** [review]
one comment, after that it looks good to me

File: e2e/protocol.test.js [hunk 6]
```diff
@@ -335,6 +335,32 @@ describe('executeScript', () => {
...
+            right: '1in',
+            bottom: '1in',
+            left: '1in'
+        },
+        preferCSSPageSize: false,
+        pageRanges: ''
+    }
+
+    it('expect string', async () => {
+        const stringData = await browser.printPage(opt)
```
Comment (christian-bromann) [inline e2e/protocol.test.js:359]:
```suggestion
    it('expect string', async () => {
        const stringData = await browser.printPage(
            'horizontal',
            1,
            false,
            '8in',
            '11in',
            '1in',
            '1in',
            '1in',
            '1in',
            false
        )
```

**Comment by Maxgit3** [comment]
@christian-bromann sorry, what was the comment?

**Comment by christian-bromann** [comment]
@Maxgit3 can you update the [protocol parameter types](https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-protocols/protocols/webdriver.json#L207-L262) to accept `string|number`. Also it would be nice if we could type the parameters for the devtools command.

**Comment by Maxgit3** [comment]
@christian-bromann did you mean add the parameters in the cli?

**Comment by christian-bromann** [comment]
> did you mean add the parameters in the cli?

no, in the protocol file that I've linked.

**Comment by Maxgit3** [comment]
aww ok thanks

**Comment by Maxgit3** [comment]
> @Maxgit3 can you update the [protocol parameter types](https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-protocols/protocols/webdriver.json#L207-L262) to accept `string|number`. Also it would be nice if we could type the parameters for the devtools command.

I am not sure what you mean by type parameters in the devtools command.  should I update the printPage.ts

File: packages/devtools/src/commands/printPage.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ * The print functions are a mechanism to render the document to a paginated format.
+ *
+ * @alias browser.printPage
+ * @see https://w3c.github.io/webdriver/#print
+ * @param {object} options an object of options that affect the pdf output
+ */
+
+import DevToolsDriver from '../devtoolsdriver'
+
+export default async function printPage(this: DevToolsDriver, options: any) {
```
Comment (christian-bromann) [inline packages/devtools/src/commands/printPage.ts:11]: The options here are `any` but we should set proper types for it.

**Review by christian-bromann — COMMENTED** [review]
Yes, see my comment below

**Comment by Maxgit3** [comment]
@christian-bromann updated file.  the scale option errors with string type

File: packages/devtools/src/commands/printPage.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        width: width,
+        height: height,
+        margin: {
+            top: top,
+            right: right,
+            bottom: bottom,
+            left: left
+        },
+        preferCSSPageSize: !shrinkToFit,
+        pageRanges: pageRanges
```
Comment (christian-bromann) [inline packages/devtools/src/commands/printPage.ts:39]:
```suggestion
        landscape: orientation === 'landscape',
        scale,
        printBackground: background,
        width,
        height,
        margin: { top, right, bottom, left },
        preferCSSPageSize: !shrinkToFit,
        pageRanges
```

**Comment by christian-bromann** [comment]
@Maxgit3 if you can transform this into a PR and sign the CLA we can get this merged.

**Comment by christian-bromann** [comment]
ping @Maxgit3

**Comment by Maxgit3** [comment]
@christian-bromann sorry about the delay

**Review by christian-bromann — APPROVED** [review]
👍

**Pull request merged**